Make unit-tests run with @woocommerce packages

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

It a workaround of https://github.com/woocommerce/woocommerce-admin/issues/6483 to make the unit-tests of this project can be run with `@woocommerce` packages

- Downgrade `@wordpress/date` to ver 3.9.0 to avoid unknown dependency conflict when running unit-tests
- Transpile all `@woocommerce` packages for jest
- Mock SVG and SCSS files for jest
- Make the pattern to be more precise when matching `.~/` alias

### Screenshots:

![image](https://user-images.githubusercontent.com/17420811/119127885-79a22780-ba67-11eb-816b-20e48c4de20f.png)

### Detailed test instructions:

1. `npm install`
2. `npm run test-unit`
3. All tests should pass

### Technical notes

#### 💡 Why this solution?

I take this opiton due to I think it would not affect this project to downgrade `@wordpress/date`. The only difference between **3.9.0** and **3.12.0** is the default `abbr` in the `timezone` setting:
- [3.9.0](https://github.com/WordPress/gutenberg/blob/%40wordpress/date%403.9.0/packages/date/src/index.js#L83)
- [3.12.0](https://github.com/WordPress/gutenberg/blob/%40wordpress/date%403.12.0/packages/date/src/index.js#L83)
- The usage of `timezone.abbr` - [date-time/timezone of @wordpress/components](https://github.com/WordPress/gutenberg/blob/%40wordpress/components%4012.0.8/packages/components/src/date-time/timezone.js)

#### 💡 The root cause of https://github.com/woocommerce/woocommerce-admin/issues/6483

I think that should be the root cause. But it needs much more time to investigate how to fix it without unexpected effects
- The result of `get( target, [ 'file', 'request' ] ) === targetName` condition is always `false` at here https://github.com/woocommerce/woocommerce-admin/blob/v2.1.2/bin/packages/get-babel-config.js#L16
- `modules` would never be set to `'commonjs'` at here https://github.com/woocommerce/woocommerce-admin/blob/v2.1.2/bin/packages/get-babel-config.js#L26-L28

### Additional note

- The demo test case https://github.com/woocommerce/google-listings-and-ads/commit/86737726add82700b5b250d76969bde68eb95412 should be reverted first once this PR be able to merge

### Other solutions 🤔 

1. Force override the source of `@wordpress/date` when resolving dependency of jest
    - I also push another solution at `fix/unit-test-with-wc-admin-ver-upgrade` branch - [diff link](https://github.com/woocommerce/google-listings-and-ads/compare/fix/unit-test-with-wc-admin-ver-upgrade)
    - The main differences to this PR are this commit https://github.com/woocommerce/google-listings-and-ads/commit/e2b26a15d6a38dfe4ff15dbc77a5cd0a4d627b3f and no need to downgrade `@wordpress/date` from the project level. But this solution looks more unstable to me
2. Rebuild all `@woocommerce` packages to CommonJS module within `node_modules` locally
    - Have copies of the below files in this project. Modify and fix them to rebuild
        - https://github.com/woocommerce/woocommerce-admin/blob/v2.1.2/bin/packages/build.js
        - https://github.com/woocommerce/woocommerce-admin/blob/v2.1.2/bin/packages/get-babel-config.js
        - https://github.com/woocommerce/woocommerce-admin/blob/v2.1.2/bin/packages/get-packages.js
    - Use [resolver](https://jestjs.io/docs/25.x/configuration#resolver-string) to point these packages to rebuilt files
    - I did an experiment to verify that. This way doesn't need to use `transformIgnorePatterns`, but I think it's too overkill. And there are still `@woocommerce` packages inside the nested `node_modules` that need to be found one by one, which looks too unrealistic to me.

### Changelog Note:

> Make unit-tests run with `@woocommerce` packages

eason9487 committed:
Downgrade @wordpress/date to ver 3.9.0

eason9487 committed:
Transpile all WC admin packages when running unit-tests

eason9487 committed:
Mock svg and scss files when running unit-tests

eason9487 committed:
Make the pattern to be more precise when matching  alias

eason9487 committed:
Add a test case for demo this fix

tomalec reviewed (approved):
Tested locally, reviewed the code. Works like charm! LGTM
Thanks for your investigation and explanations.

tomalec committed:
Move `jest.config` code comment to the matching line

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/649/files#r638079878
